File: api/src/main/kotlin/org/team619/rutor/model/Caption.java
package org.team619.rutor.model;

import org.team619.rutor.model.core.DefaultEntity;

import java.io.File;

/**
 * Created by BORIS on 07.08.2016.
 */
public final class Caption implements DefaultEntity {

    private final String adaptedName;
    private final String name;
    private final String subtitle;
    private final String year;

    public Caption(String title, String year, String subtitle) {
        this.adaptedName = parse(title, 0);
        this.name = parse(title, 1);
        this.subtitle = subtitle;
        this.year = year;
    }

    private String parse(String title, int i) {
        return (title != null && title.contains(File.separator)) ? title.split(File.separator)[i] : title;
    }

    public Caption(String title) {
        this.adaptedName = null;
        this.name = title;
        this.subtitle = null;
        this.year = null;
    }

    public String getAdaptedName() {
        return adaptedName;
    }

    public String getName() {
        return name;
    }

    public String getTitle() {
        if (adaptedName == null) {
            return name;
        } else {
            return adaptedName + File.separator + name;
        }
    }

    public String getSubtitle() {
        return subtitle;
    }

    public String getYear() {
        return year;
    }

}

File: api/src/main/kotlin/org/team619/rutor/model/Caption.kt
package org.team619.rutor.model

import org.team619.rutor.core.Page

/**
 * Created by BORIS on 07.08.2016.
 */
class Caption : Page {

    val adaptedName: String
    val name: String
    val subtitle: String
    val year: String
    val title: String
        get() = if (adaptedName == null) name else adaptedName + "/" + name

    constructor(name: String) {
        this.adaptedName = ""
        this.name = name
        this.subtitle = ""
        this.year = ""
    }

    constructor(title: String, year: String, subtitle: String) {
        this.adaptedName = parse(title, 0)
        this.name = parse(title, 1)
        this.subtitle = subtitle
        this.year = year
    }

    constructor(adaptedName: String, name: String, year: String, subtitle: String) {
        this.adaptedName = adaptedName
        this.name = name
        this.subtitle = subtitle
        this.year = year
    }

    private fun parse(title: String?, i: Int): String {
        return java.lang.String.valueOf((if (title != null && title.contains("/")) title.split("/".toRegex())[i] else title)?.trim())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Caption

        if (adaptedName != other.adaptedName) return false
        if (name != other.name) return false
        if (subtitle != other.subtitle) return false
        if (year != other.year) return false

        return true
    }

    override fun hashCode(): Int {
        var result = adaptedName.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + subtitle.hashCode()
        result = 31 * result + year.hashCode()
        return result
    }

    override fun toString(): String {
        return "Caption(adaptedName=$adaptedName, name=$name, subtitle=$subtitle, year=$year)"
    }

    companion object {
        fun empty(): Caption = Caption("", "", "")
    }

}
